Source/WebCore/platform/graphics/BitmapImageSource.cpp

```diff
@@ -0,0 +1,484 @@
+
+    auto buffer = ImageBuffer::create(size, RenderingPurpose::Unspecified, 1, DestinationColorSpace::SRGB(), PixelFormat::BGRA8);
+    if (!buffer)
+        return nativeImage;
+
+    auto destinationRect = FloatRect { FloatPoint(), size };
+    auto sourceRect = FloatRect { FloatPoint(), sourceSize };
+
+    buffer->context().drawNativeImage(*nativeImage, destinationRect, sourceRect, { orientation });
+    return ImageBuffer::sinkIntoNativeImage(WTFMove(buffer));
+}
+
+Expected<Ref<NativeImage>, DecodingStatus> BitmapImageSource::nativeImageAtIndexForDrawing(unsigned index, SubsamplingLevel subsamplingLevel, const DecodingOptions& options)
+{
+    if (auto frameProvider = ensureFrameProvider())
+        return frameProvider->nativeImageAtIndexForDrawing(index, subsamplingLevel, options);
+    return makeUnexpected(DecodingStatus::Invalid);
+}
+
+unsigned BitmapImageSource::currentFrameIndex() const
+{
+    return m_animator ? m_animator->currentFrameIndex() : primaryFrameIndex();
+}
+
+RefPtr<NativeImage> BitmapImageSource::currentPreTransformedNativeImage(ImageOrientation orientation)
+{
+    return preTransformedNativeImageAtIndex(currentFrameIndex(), orientation);
+}
+
+Expected<Ref<NativeImage>, DecodingStatus> BitmapImageSource::currentNativeImageForDrawing(SubsamplingLevel subsamplingLevel, const DecodingOptions& options)
+{
+    startAnimation(subsamplingLevel, options);
+    return nativeImageAtIndexForDrawing(currentFrameIndex(), subsamplingLevel, options);
+}
+
+IntSize BitmapImageSource::size(ImageOrientation orientation) const
+{
+    auto densityCorrectedSize = m_metaData.densityCorrectedSize();
+    if (!densityCorrectedSize)
+        return sourceSize(orientation);
+
+    if (orientation == ImageOrientation::Orientation::FromImage)
+        orientation = m_metaData.orientation();
+
+    return orientation.usesWidthAsHeight() ? densityCorrectedSize->transposedSize() : *densityCorrectedSize;
+}
+
+IntSize BitmapImageSource::sourceSize(ImageOrientation orientation) const
+{
+#if !USE(CG)
+    // It's possible that we have decoded the metadata, but not frame contents yet. In that case ImageDecoder claims to
+    // have the size available, but the frame cache is empty. Return the decoder size without caching in such case.
+    if (m_decoder && !(m_frameProvider && m_frameProvider->frameCount()))
+        return m_decoder->size();
+#endif
+
+    auto size = m_metaData.size();
+
+    if (orientation == ImageOrientation::Orientation::FromImage)
+        orientation = m_metaData.orientation();
+
+    return orientation.usesWidthAsHeight() ? size.transposedSize() : size;
+}
+
+unsigned BitmapImageSource::frameCount() const
+{
+    ASSERT_IMPLIES(m_frameProvider, m_frameProvider->frameCount() == m_metaData.frameCount());
+    return m_metaData.frameCount();
+}
+
+std::optional<Color> BitmapImageSource::singlePixelSolidColor() const
+{
+    if (notSolidColor())
+        return std::nullopt;
```

Comment:
Do we need this? Won't the `SinglePixelSolidColor` metadata be set only if `NativeImage` returned a single pixel color?

Reply:
`hasSolidColor()` checks the `size()` is one pixel and the `frameCount()` is one frame. `BitmapImageSource::singlePixelSolidColor()` should return a Color if `BitmapImage` can always be displayed by filling rect with one single color. If it is an animated image, this function should return std::nullopt even if the primary frame is a single color.

Reply:
I see, so it's checking for an animated image that is the important part here.